trieste/acquisition/interface.py

```diff
@@ -49,13 +49,16 @@ def __call__(self, x: TensorType) -> TensorType:
         """Call acquisition function."""
 
 
-class AcquisitionFunctionBuilder(ABC):
+T = TypeVar("T", bound=ProbabilisticModel)
+
+
+class AcquisitionFunctionBuilder(Generic[T], ABC):
```

Comment:
a bit of education please :)

what does adding this `Generic[T]` bit do for the class?

Reply:
"This PR enables mypy to enforce that the correct model types are used when constructing the AcquistionFunction." this would happen only in tests, no?

Reply:
No problem :)

This basically makes it possible for classes inheriting from `AcquisitionFunctionBuilder` to specify that they require specific subclasses of `ProbabilisticModel` (i.e. rather than just the base `ProbabilisticModel` itself), and that that type should be consistent wherever you see the symbol `T`. 

For example, imagine if you have `SpecialProbabilisticModel`, which is a subclass that extends `ProbabilisticModel` with extra functionality required by `SpecialAcquisitionFunctionBuilder`. You want to make sure that nobody passes a standard `ProbabilisticModel` into `SpecialAcquisitionFunctionBuilder`'s `prepare_acquisition_function` (because it doesn't have the required extra functionality), by declaring `SpecialAcquisitionFunctionBuilder` as `SpecialAcquisitionFunctionBuilder[SpecialProbabilisticModel]` mypy can enforce that is true. If `AcquisitionFunctionBuilder` isn't generic, everything _seems_ ok (because everything is a `ProbabilisticModel`), until a runtime error occurs.

Reply:
thanks a lot, that's much clearer and quite useful, we were thinking of how to something along those lines

some follow up questions - how can mypy enforce that? only in tests we actually pass some models to acq functions, right? mypy wouldn't do anything for the end-user, if the user ignores the documentation where the types would be visible it would still be the runtime error

Reply:
If the end user uses mypy (which is always a good idea, though academic users probably won't) then it would catch it. Otherwise it would fail with a (possibly confusing) runtime error. I expect we'll continue to check the types at runtime for this reason (so we can give a clear error message) though it would be good to give precise types too — not only does it help people who use mypy, it also serves as better documentation for the acquisition functions.